HTTPlayground/views.py

```diff
@@ -1,50 +1,59 @@
     send_headers(request, content_type='text/html', **cookie_header)
-    t = Template(FileReader.read('login.html'))
-    rendered_html = t.render(link=f'{ACCESSORY_URL_PREFIX}/form')
+    file = TEMPLATE_ENV.get_template('base_response_form.html')
+    rendered_html = file.render(title='Login', link=f'{ACCESSORY_URL_PREFIX}/form', text='Successful login.')
     return rendered_html
 
 
+@require_http_methods(['GET'], not_found)
 def logout(request):
-    cookie_header = {'Set-Cookie': CookieHandler.generate_cookie(request, {'auth': 'False'})}
+    cookie_header = {'Set-Cookie': CookieHandler.generate_cookie_for_remove(request, name='auth')}
     send_headers(request, content_type='text/html', **cookie_header)
-    t = Template(FileReader.read('logout.html'))
-    rendered_html = t.render(link=f'{ACCESSORY_URL_PREFIX}/form')
+    file = TEMPLATE_ENV.get_template('base_response_form.html')
+    rendered_html = file.render(title='Logout', link=f'{ACCESSORY_URL_PREFIX}/form', text='Successful logout.')
     return rendered_html
 
 
+@require_http_methods(['GET'], not_found)
 def get_form(request):
     send_headers(request, content_type='text/html')
-    form_text = {'form_link': (f'{ACCESSORY_URL_PREFIX}/charge', f'{ACCESSORY_URL_PREFIX}/charge'),
-                 'auth_link': (f'{ACCESSORY_URL_PREFIX}/authh/login', f'{ACCESSORY_URL_PREFIX}/authh/logout'),
-                 'auth_text': ('Login', 'Logout')}
-    is_authenticated = CookieHandler.is_cookie(request, {'auth': 'True'})
-    t = Template(FileReader.read('form.html'))
-    rendered_html = t.render(**{k: v[is_authenticated] for k, v in form_text.items()})
+    form_text = {'title': 'form',
+                 'form_link': f'{ACCESSORY_URL_PREFIX}/charge',
+                 'auth_link': f'{ACCESSORY_URL_PREFIX}/authh/login',
+                 'deauth_link': f'{ACCESSORY_URL_PREFIX}/authh/logout',
+                 'auth_text': 'Login',
+                 'deauth_text': 'Logout'}
+    file = TEMPLATE_ENV.get_template('form.html')
+    rendered_html = file.render(**form_text)
     return rendered_html
 
 
+@require_http_methods(['POST'], not_found)
 def process_form(request):
     send_headers(request, content_type='text/html')
-    if not CookieHandler.is_cookie(request, {'auth': 'True'}):
-        return
+    file = TEMPLATE_ENV.get_template('base_response_form.html')
+    if not CookieHandler.has_cookie(request, name='auth', value='True'):
+        return file.render(title='Error', link=f'{ACCESSORY_URL_PREFIX}/form', text='Not authorized!')
     form = FieldStorage(
         fp=request.rfile,
         headers=request.headers,
         environ={'REQUEST_METHOD': 'POST'}
     )
-    t = Template(FileReader.read('charged.html'))
-    rendered_html = t.render(link=f'{ACCESSORY_URL_PREFIX}/form', text=f'OK. {form.getvalue("sum_value")}$ charged!')
+    rendered_html = file.render(title='Charged', link=f'{ACCESSORY_URL_PREFIX}/form',
```

Comment:
What happens if the cookie is fine, but the origin is incorrect (not assumed to have access here)? 
Seems like:
1. money will be charged anyways 
1. UI of the user (on origin) will break with no sensible message

Correct me if I'm wrong, but otherwise both problems are quite serious (the first one is even critical).

Reply:
CSRF protection in task4.

> What happens if the cookie is fine, but the origin is incorrect (not assumed to have access here)?
> Seems like:
> 
>     1. money will be charged anyways
> 
>     2. UI of the user (on origin) will break with no sensible message
> 
> 
> Correct me if I'm wrong, but otherwise both problems are quite serious (the first one is even critical).